nimbess-cni.go

```diff
@@ -0,0 +1,314 @@
+// Copyright (c) 2019 Red Hat and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Main package for the Nimbess CNI plugin
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"time"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/version"
+	"google.golang.org/grpc"
+
+	cnitypes "github.com/containernetworking/cni/pkg/types/current"
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+	cninimbess "github.com/nimbess/nimbess-agent/pkg/proto/cni"
+	log "github.com/sirupsen/logrus"
+)
+
+const defaultLogFile = "/tmp/nimbess-cni.log"
+
+// NimbessConfig is whatever you expect your configuration json to be. This is whatever
+// is passed in on stdin. Your plugin may wish to expose its functionality via
+// runtime args, see CONVENTIONS.md in the CNI spec.
+type nimbessConfig struct {
+	// common CNI config
+	types.NetConf
+
+	// PrevResult contains previous plugin's result, used only when called in the context of a chained plugin.
+	PrevResult *map[string]interface{} `json:"prevResult"`
+
+	// GrpcServer is a plugin-specific config, contains location of the gRPC server
```

Comment:
ah, ok so guess this answers my question about common... there are different plugins each with unique config. Are there more than one plugins?

Reply:
CNI is just a specification:

https://github.com/containernetworking/cni/blob/master/SPEC.md

The gRPC config which points to a server started up by the agent, does not belong to that spec, while things like container id, namespace, etc, do.

Reply:
Thanks for clarification. When I saw common and then plugin-specific I was thinking MULTIPLE plugins that were of different types. They all had a common section and then plugin-specific sections for each of the different plugins. But I should have read this as the common section is for the CNI spec and then our plugin here has specific pieces.